nemo/collections/asr/parts/k2/topologies.py

```diff
@@ -0,0 +1,136 @@
+    elif name == "minimal":
+        return build_minimal_topo(tokens)
+    else:
+        raise ValueError(f"Unknown topo name: {name}")
+
+
+def build_default_topo(tokens: List[int], with_selfloops: bool = True) -> k2.Fsa:
+    """Build the default CTC topology.
+    """
+    assert 0 in tokens, "We assume 0 is ID of the blank symbol"
+
+    num_states = len(tokens)
+    final_state = num_states
+    arcs = "" if with_selfloops else "0 0 0 0 0.0\n"
+    for i in range(num_states):
+        for j in range(num_states):
+            if i == j:
+                if with_selfloops:
+                    arcs += f"{i} {i} {tokens[i]} 0 0.0\n"
+            else:
+                arcs += f"{i} {j} {tokens[j]} {tokens[j]} 0.0\n"
+        arcs += f"{i} {final_state} -1 -1 0.0\n"
+    arcs += f"{final_state}"
+    ans = k2.Fsa.from_str(arcs, num_aux_labels=1)
+    ans = k2.arc_sort(ans)
+    return ans
+
+
+def build_compact_topo(tokens: List[int], with_selfloops: bool = True) -> k2.Fsa:
+    """Build the compact CTC topology.
+    See https://arxiv.org/abs/2110.03098
+    """
+    assert 0 in tokens, "We assume 0 is ID of the blank symbol"
+
+    selfloops_shift = int(with_selfloops)
+    blank_num = 1
+    num_states = len(tokens) + selfloops_shift
+    final_state = num_states
+    arcs = f"0 {selfloops_shift} {blank_num} 0 0.0\n"
+    for i in range(blank_num + selfloops_shift, num_states):
+        arcs += f"0 {i} {tokens[i - selfloops_shift] + 1} {tokens[i - selfloops_shift] + 1} 0.0\n"
+    arcs += f"0 {final_state} -1 -1 0.0\n"
+    for i in range(blank_num, num_states):
+        arcs += f"{i} 0 0 0 0.0\n"
+        if with_selfloops:
+            arcs += f"{i} {i} {tokens[i - selfloops_shift] + 1} 0 0.0\n"
+    arcs += f"{final_state}"
+    ans = k2.Fsa.from_str(arcs, num_aux_labels=1)
+    ans = k2.arc_sort(ans)
+    return ans
+
+
+def build_shared_blank_topo(tokens: List[int], with_selfloops: bool = True) -> k2.Fsa:
+    """Build the shared blank CTC topology.
+    See https://github.com/k2-fsa/k2/issues/746#issuecomment-856421616
+    """
+    assert 0 in tokens, "We assume 0 is the ID of the blank symbol"
+
+    tokens = tokens.copy()
+    tokens.remove(0)
+    num_tokens = len(tokens)
+    start = 0
+    final = num_tokens + 1
+    arcs = []
+    arcs.append([start, start, 0, 0, 0])
+    arcs.append([start, final, -1, -1, 0])
+    arcs.append([final])
+    for i, p in enumerate(tokens):
+        i += 1
+        arcs.append([start, start, p, p, 0])
+        arcs.append([start, i, p, p, 0])
+        arcs.append([i, start, p, 0, 0])
+        if with_selfloops:
+            arcs.append([i, i, p, 0, 0])
+    arcs = sorted(arcs, key=lambda arc: arc[0])
+    arcs = [[str(i) for i in arc] for arc in arcs]
+    arcs = [" ".join(arc) for arc in arcs]
+    arcs = "\n".join(arcs)
+    ans = k2.Fsa.from_str(arcs, num_aux_labels=1)
+    ans = k2.arc_sort(ans)
+    return ans
+
+
+def build_minimal_topo(tokens: List[int]) -> k2.Fsa:
```

Comment:
Same as above

Reply:
Done in [7b2b4a4](https://github.com/NVIDIA/NeMo/pull/3351/commits/7b2b4a4ba4e34848d526d8c4ac53903bfbfe2fd2)